lib/vtls/schannel.c

```diff
@@ -405,6 +405,362 @@ get_cert_location(TCHAR *path, DWORD *store_name, TCHAR **store_path,
+        CERT_FIND_HASH, &cert_thumbprint, NULL);
+
+      curlx_unicodefree(cert_path);
+
+      if(client_certs[0]) {
+        schannel_cred.cCreds = 1;
+        schannel_cred.paCred = client_certs;
+      }
+      else {
+        /* CRYPT_E_NOT_FOUND / E_INVALIDARG */
+        CertCloseStore(cert_store, 0);
+        return CURLE_SSL_CERTPROBLEM;
+      }
+    }
+    CertCloseStore(cert_store, 0);
+  }
+  else {
+    schannel_cred.dwFlags |= SCH_CRED_NO_DEFAULT_CREDS;
+  }
+#else
+  if(data->set.ssl.primary.clientcert) {
+    failf(data, "schannel: client cert support not built in");
+    return CURLE_NOT_BUILT_IN;
+  }
+#endif
+
+  /* allocate memory for the re-usable credential handle */
+  BACKEND->cred = (struct Curl_schannel_cred *)
+    calloc(1, sizeof(struct Curl_schannel_cred));
+  if(!BACKEND->cred) {
+    failf(data, "schannel: unable to allocate memory");
+
+    if(client_certs[0])
+      CertFreeCertificateContext(client_certs[0]);
+
+    return CURLE_OUT_OF_MEMORY;
+  }
+  BACKEND->cred->refcount = 1;
+
+  sspi_status =
+    s_pSecFn->AcquireCredentialsHandle(NULL, (TCHAR *)UNISP_NAME,
+                                       SECPKG_CRED_OUTBOUND, NULL,
+                                       &schannel_cred, NULL, NULL,
+                                       &BACKEND->cred->cred_handle,
+                                       &BACKEND->cred->time_stamp);
+
+  if(client_certs[0])
+    CertFreeCertificateContext(client_certs[0]);
+
+  if(sspi_status != SEC_E_OK) {
+    char buffer[STRERROR_LEN];
+    failf(data, "schannel: AcquireCredentialsHandle failed: %s",
+          Curl_sspi_strerror(sspi_status, buffer, sizeof(buffer)));
+    Curl_safefree(BACKEND->cred);
+    switch(sspi_status) {
+      case SEC_E_INSUFFICIENT_MEMORY:
+        return CURLE_OUT_OF_MEMORY;
+      case SEC_E_NO_CREDENTIALS:
+      case SEC_E_SECPKG_NOT_FOUND:
+      case SEC_E_NOT_OWNER:
+      case SEC_E_UNKNOWN_CREDENTIALS:
+      case SEC_E_INTERNAL_ERROR:
+      default:
+        return CURLE_SSL_CONNECT_ERROR;
+    }
+  }
+
+  return CURLE_OK;
+}
+
+static void schannel_session_free(void *ptr);
+
+static CURLcode
+schannel_client_certificate(struct Curl_easy *data, struct connectdata *conn,
+                            int sockindex)
+{
+  struct ssl_connect_data *connssl = &conn->ssl[sockindex];
+  SECURITY_STATUS status = 0;
+  SecPkgContext_IssuerListInfoEx dist_names;
+  CURLcode result;
+  unsigned long flags = BACKEND->ret_flags;
+
+  infof(data, "schannel: a client certificate has been requested\n");
+
+  /* Retrieve the distinguished names */
+  status =
+    s_pSecFn->QueryContextAttributes(&BACKEND->ctxt->ctxt_handle,
+                                     SECPKG_ATTR_ISSUER_LIST_EX,
+                                     &dist_names);
+  if(status != SEC_E_OK) {
+    infof(data, "QueryContextAttributes failed. status = %d, %ul ", status,
+          flags);
+    return CURLE_SSL_CONNECT_ERROR;
+  }
```

Comment:
`else {}` block is unnecessary here since we just return in above block.

Reply:
removed the else.

Reply:
Again the code already existed, it just moved to another function.